lib/Parse/SyntaxParsingCache.cpp

```diff
@@ -85,7 +85,7 @@ llvm::Optional<Syntax> SyntaxParsingCache::lookUp(size_t NewPosition,
   size_t OldPosition = NewPosition;
   for (auto I = Edits.rbegin(), E = Edits.rend(); I != E; ++I) {
     auto Edit = *I;
-    if (Edit.End <= OldPosition) {
+    if (Edit.End < OldPosition) {
```

Comment:
After some more consideration this should be 
```cpp
if (Edit.End + Edit.ReplacementLength - Edit.originalLength() <= OldPosition) {
```

## Explanation

With `<=` we don't catch the following case:

Original text: `foobar`
Change `foo` to `fo` resulting in `fobar`

The edit range is 1-4 with original length 3 and replacement length 2. 
If I now try to look up `bar`, we start with `OldPosition = NewPosition = 3`. Now we need to see if that position needs to be adjusted because of the edit (which it obviously needs to be), but we get `Edit.End (4) <= OldPosition (3)` does not hold. 

The issue is that we are comparing `OldPosition` which is still in the position space of the post-edit file and `Edit.End` which is in the position space of the pre-edit file, hence we need to translate `Edit.End` to the post-position space as well.

Reply:
Ah, you're right! Ok, so this should make it correct for a single edit, but if there are multiple I think we'll need to account for the original-vs-replacement length differences of any edits earlier in the file too. I'll give that a go and update shortly.